### ISSUE TITLE
linter: eslint/prefer-const warns on deferred assignment closure pattern

### ISSUE THREAD

**Issue by howmanysmall** [oxc-project/oxc#19654]
### What version of Oxlint are you using?

1.49.0

### What command did you run?

```bash
oxlint -D prefer-const src/shared/meta/oxlint-prefer-const-repro.ts
```

### What does your `.oxlintrc.json` (or `oxlint.config.ts`) config file look like?

No config file for this repro.

### What happened?

Oxlint reports:

```text
x eslint(prefer-const): `controllerApiReference` is never reassigned.
help: Use `const` instead.
```

Repro code:

```ts
interface ControllerApi {
	pause: () => void;
	start: () => void;
	stop: () => void;
}

function createControllerApi(): ControllerApi {
	let controllerApiReference: ControllerApi | undefined;

	function pause(): void {
		resolveController();
	}

	function resolveController(): ControllerApi | undefined {
		if (controllerApiReference === undefined) return undefined;
		return controllerApiReference;
	}

	function start(): void {
		resolveController();
	}

	function stop(): void {
		resolveController();
	}

	const controllerApi: ControllerApi = { pause, start, stop };

	controllerApiReference = controllerApi;
	return controllerApi;
}

createControllerApi();
```

`controllerApiReference` is declared first so the inner functions can close over it, then it's assigned once after the API object exists (basically a self-reference without relying on `this`).

The suggestion to use `const` isn't actionable as a simple change here, since `const` needs an initializer and there isn't one at the declaration site.

Expected: no `prefer-const` diagnostic for this pattern (or at least don't suggest `const` when there is no initializer to attach it to).

**Comment by connorshea** [comment]
[This occurs in ESLint as well, it looks like](https://eslint.org/play/#eyJ0ZXh0IjoiaW50ZXJmYWNlIENvbnRyb2xsZXJBcGkge1xuXHRwYXVzZTogKCkgPT4gdm9pZDtcblx0c3RhcnQ6ICgpID0+IHZvaWQ7XG5cdHN0b3A6ICgpID0+IHZvaWQ7XG59XG5cbmZ1bmN0aW9uIGNyZWF0ZUNvbnRyb2xsZXJBcGkoKTogQ29udHJvbGxlckFwaSB7XG5cdGxldCBjb250cm9sbGVyQXBpUmVmZXJlbmNlOiBDb250cm9sbGVyQXBpIHwgdW5kZWZpbmVkO1xuXG5cdGZ1bmN0aW9uIHBhdXNlKCk6IHZvaWQge1xuXHRcdHJlc29sdmVDb250cm9sbGVyKCk7XG5cdH1cblxuXHRmdW5jdGlvbiByZXNvbHZlQ29udHJvbGxlcigpOiBDb250cm9sbGVyQXBpIHwgdW5kZWZpbmVkIHtcblx0XHRpZiAoY29udHJvbGxlckFwaVJlZmVyZW5jZSA9PT0gdW5kZWZpbmVkKSByZXR1cm4gdW5kZWZpbmVkO1xuXHRcdHJldHVybiBjb250cm9sbGVyQXBpUmVmZXJlbmNlO1xuXHR9XG5cblx0ZnVuY3Rpb24gc3RhcnQoKTogdm9pZCB7XG5cdFx0cmVzb2x2ZUNvbnRyb2xsZXIoKTtcblx0fVxuXG5cdGZ1bmN0aW9uIHN0b3AoKTogdm9pZCB7XG5cdFx0cmVzb2x2ZUNvbnRyb2xsZXIoKTtcblx0fVxuXG5cdGNvbnN0IGNvbnRyb2xsZXJBcGk6IENvbnRyb2xsZXJBcGkgPSB7IHBhdXNlLCBzdGFydCwgc3RvcCB9O1xuXG5cdGNvbnRyb2xsZXJBcGlSZWZlcmVuY2UgPSBjb250cm9sbGVyQXBpO1xuXHRyZXR1cm4gY29udHJvbGxlckFwaTtcbn1cblxuY3JlYXRlQ29udHJvbGxlckFwaSgpOyIsIm9wdGlvbnMiOnsicnVsZXMiOnsicHJlZmVyLWNvbnN0IjpbImVycm9yIl19LCJsYW5ndWFnZU9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hRmVhdHVyZXMiOnsianN4Ijp0cnVlfSwic291cmNlVHlwZSI6Im1vZHVsZSJ9LCJlY21hVmVyc2lvbiI6ImxhdGVzdCIsInBhcnNlciI6IkB0eXBlc2NyaXB0LWVzbGludC9wYXJzZXIifX19)

**Comment by howmanysmall** [comment]
Nailed the compatibility then 😭

**Comment by camc314** [comment]
Closing as not planned - as connor mentioned, we have the same behaviour as eslint - I would just disable the rule for this specific case or open an issue upstream and see what the eslint team says.
